### ISSUE TITLE
Donation of payment method abstraction

### ISSUE THREAD

**Issue by svallory** [sharetribe/sharetribe#1395]
Hi, I've been hired to implement a new payment method (namely https://moip.com.br/). The way I want to go about it is to abstract away the payment method using either Strategy pattern, configured via `config.yml` or a plugin structure where many methods could be available at the same time. Then I would make a pull request to integrate it to the main project. That is considerable more work than I have to do, but I think you guys deserve the contribution for the great work you have done.

What I would like to know is:

1. Is it of interest to you? (I don't want to do the extra work unless it's going to make it to the main repository)

2. Can someone spend an hour or 2 to help expedite the development by showing me around the payment workflow?

That's it. I hope you guys like the idea. I'm pretty excited to work on this.

**Comment by rap1ds** [comment]
Thank you so much for you initiative! We greatly appreciate it!

TL;DR, Adding country specific gateways is definitely something we want to do, but now right now. This is not high on our roadmap, so that's why we will not merge it to the master.

Longer story:

As I mentioned, adding country specific gateways is something we are very interested and something we feel that the OS community could help us to do. However, as discussed, it would require some kind of clear API or plugin architecture which we do not have at the moment. I feel that implementing the plugin architecture, even if it's done by you, would also require extensive time from the core team (helping you to understand our not so simple app and code structure, code reviews, etc.) and at the moment we can not guarantee we have enough time for that.

This may sound pretty silly, you know, someone is donating us free work, and we are basically saying no to it. However, all the code that we add to the application raises the level of complexity of the application, and we're not ready for doing that for features that are not high on the roadmap. In addition, even if the payment gateway integration is done by you, in the end the responsibility for maintaining it would be on core team's shoulders. I hope you understand :)

Anyway, I'd be really happy to see what you are building, so if you are able to share the repository of your project, I'd like to see how the payment integration works!

**Comment by svallory** [comment]
I understand those points. I've been there, so I really do. But I don't think what I proposed is clear enough.

Forget about moip, what I'm offering is: to abstract the **current** payment structure in such a way that it's easier to add new gateways. The end result would be that paypal and braintree gateways are used by the rest of the code through a common interface. The only code added to sharetribe would be the code which selects which gateway to use based on configuration and the abstraction layer itself. It's more or less just a refactoring of what you have now.

There are two strategies to achieving that: a plugin architecture (with methods for detecting available plugins, loading and bootstrapping  each one) or the [Strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern) (which is simpler but more "entangled" with the code). 

A plugin architecture would make it easier for the community to contribute with other gateways, which could be installed using gems but it's a lot more work defining (and that would need your thought and approval. A simple Strategy pattern still allows contribution, gateways could live in separate projects, but the installation would have to be done manually. 

As for the time you'd spend on this, that depends on the strategy you choose. I'm not counting on guidance throughout the development. I got paypal working in less than 24hs, so I think I'll be fine doing it by myself. But it sure would be nice to be properly introduced to the payment workflow :)

**Comment by svallory** [comment]
@rap1ds any thoughts?

**Comment by rap1ds** [comment]
Sorry, I may have been a little unclear with my answer. I did understand that the main point is the abstraction of the payment gateways, not Moip integration.

Sure, if you want to build it, go ahead! 😊 But what I'm saying here that I can not promise that it will end up to the Sharetribe master branch. In fact, I expect that it will not be merged.

And sorry about my skeptic attitude. I can explain you a little bit why I am a little bit skeptic about this.

Payment gateway integrations in a multivendor context like Sharetribe (multiple sellers, multiple buyers) may be extremely difficult to implement. Braintree integration was rather easy, I guess Stripe integration would be also pretty easy, but boy it was hard to implement the Express Checkout with PayPal.

The abstraction layer will be as complex as the most complex gateway integration (PayPal). I guess it would be a lot easier to do such abstraction to cover only for example Braintree and Stripe, but you have to make the abstraction so that also PayPal works through it flawlessly.

So here are some things the abstraction layer has to support:

- Payer registration: Braintree requires data like name, address, account number etc, where as PayPal requires us to redirect user to PayPal.com and do a couple of API calls. After that the PayPal user is redirected back to our site and in a separate redirect we have to redirect the user back to PayPal in order to ask for the Billing Agreement. Both Braintree and PayPal has separate view/controllers and database tables for this process, so that API should be handle those both cases.
- Payment: Braintree requires Credit Card information and one single API call, PayPal requires the "pay with PayPal button", a couple API calls and redirecting user to PayPal and then back to Sharetribe again. Both have their own views and controllers at the moment.
- Webhooks: After payer registration, Braintree does some manual checking that the payer is ok. When that is done, they send us a webhook notification that the registration was successful and the account is ready to be used. PayPal uses "IPN" notifications. If for example during the payment when the user is redirected to PayPal to pay the user closes the browser so that she is never redirected to Sharetribe, we get a IPN after a small delay. So the abstraction layer should also at some level handle the Webhooks I guess?
- Commission calculations: Different gateways have different commissions. Those calculations should be behind the abstraction layer.
- ...and a lot more.

So, again. I believe building a abstraction layer to payment gateways is not an easy job. If it was, we would have done it already. In addition, building this kind of abstraction is quite crucial architectural part of the Sharetribe codebase and that's why I think that it's something that maybe should be done by the core team.

**Comment by svallory** [comment]
@rap1ds Thank you very much for the thoughtful response. 

I know this is not an easy task and I completely understand, and agree, that it is in fact a job for the core team. Now that I know you understood what I proposed, I'll leave it be. I don't know if I'll abstract the payment or not, since that's more work and the only incentive I had was for it to be merged, but I'll let you know how it goes.

One piece of advice though... If sometime in the future you decide to build it, remember that the abstraction layer (or the payment plugin api, if you will) is defined *only* by the application needs. Then it's the payment gateway implementation that conforms to the abstraction layer, not the other way around.

So the abstraction is not as complex as the most complex gateway. In fact, it hides gateways complexity. It will be as complex as the app requires it to be.

Anyway, thank you guys very much for the fast and detailed responses :)

**Comment by JarleKristiansen** [comment]
@svallory 
Iam setting up a open source marketplace. 
Can I hier you/get an offer from you to set up the stripe payment gateway ?
Please contact me on jarle@viewinvest.no and I can give you more details.

**Comment by delblues** [comment]
Hi @svallory ! 

I'm from Brazil and I'm looking for a solution for integrating sharetribe open source with a gateway that works here. In my research I came across this topic and It would be great if you could help me. Did you  succeeded in this integration with MOIP for example? Did you managed to develop this plugin the way you exposed above?  thank you!

**Comment by svallory** [comment]
Hi @delblues, I did part of it, but the project was abandoned (by the client) so I never finished it.

**Comment by delblues** [comment]
@svallory thanks for answer it.
